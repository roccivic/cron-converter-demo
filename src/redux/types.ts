export type CronState = {
  expression: string;
  array: number[][];
  error: string;
  prev: string;
  next: string;
};

export type ValuePayload = {
  index: number;
  values: number[];
};
