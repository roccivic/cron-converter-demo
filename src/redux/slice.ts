import { arrayToString, getSchedule, stringToArray, Schedule } from "cron-converter";
import { createSlice } from "@reduxjs/toolkit";
import { CronState, ValuePayload } from "./types";
import { DateTime } from "luxon";
import { PayloadAction } from "@reduxjs/toolkit";

const updateSchedule = (state: CronState) => {
  const schedule : Schedule = getSchedule(state.array);
  state.next = schedule.next().toLocaleString(DateTime.DATETIME_FULL);
  state.prev = schedule.prev().toLocaleString(DateTime.DATETIME_FULL);
  return state;
};

export const getInitialState = () => {
  const expression = "*/15 0-11 1,10,20 * *";
  const array = stringToArray(expression);
  return updateSchedule({
    expression,
    array,
    error: "",
    prev: "",
    next: "",
  });
};

export const slice = createSlice({
  name: "cron",
  initialState: getInitialState(),
  reducers: {
    SET_EXPRESSION: (state, action: PayloadAction<string>) => {
      state.error = "";
      state.expression = action.payload;
      try {
        state.array = stringToArray(action.payload);
        updateSchedule(state);
      } catch (e: any) {
        state.error = e.message;
      }
    },
    SET_VALUE: (state, action: PayloadAction<ValuePayload>) => {
      state.error = "";
      state.array[action.payload.index] = action.payload.values;
      try {
        const expression = arrayToString(state.array);
        state.expression = expression;
        updateSchedule(state);
      } catch (e: any) {
        state.error = e.message;
      }
    },
  },
});

export const { SET_EXPRESSION, SET_VALUE } = slice.actions;

export default slice.reducer;
