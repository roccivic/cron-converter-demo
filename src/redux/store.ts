import { configureStore } from '@reduxjs/toolkit'
import { TypedUseSelectorHook, useSelector } from 'react-redux'
import reducer from './slice'

export const store = configureStore({ reducer })

export type RootState = ReturnType<typeof store.getState>
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector

