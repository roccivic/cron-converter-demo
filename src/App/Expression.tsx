import { useDispatch } from "react-redux";
import { SET_EXPRESSION } from "../redux/slice";
import { useAppSelector } from "../redux/store";

export const Expression = () => {
  const dispatch = useDispatch();
  const value = useAppSelector((state) => state.expression);
  return (
    <div className="form-floating" style={{ marginBottom: 15 }}>
      <input
        type="email"
        className="form-control"
        id="expression"
        onChange={(e) => dispatch(SET_EXPRESSION(e.target.value))}
        value={value}
      />
      <label htmlFor="expression">Cron expression</label>
    </div>
  );
};
