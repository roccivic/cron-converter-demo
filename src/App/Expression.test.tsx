import { fireEvent, render, screen } from "@testing-library/react";
import { SET_EXPRESSION } from "../redux/slice";
import { Expression } from "./Expression";

const mockDispatch = jest.fn();

jest.mock("react-redux", () => ({
  useDispatch: () => mockDispatch,
  useSelector: jest.fn().mockImplementation(() => "EXPRESSION"),
}));

test("<Expression />", () => {
  const comp = render(<Expression />);
  expect(comp.container).toMatchSnapshot();
  fireEvent.change(screen.getByDisplayValue("EXPRESSION"), { target: { value: "NEW_VALUE" } });
  expect(mockDispatch).toHaveBeenCalledWith(SET_EXPRESSION("NEW_VALUE"));
});
