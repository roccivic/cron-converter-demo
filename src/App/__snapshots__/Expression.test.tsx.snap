// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Expression /> 1`] = `
<div>
  <div
    class="form-floating"
    style="margin-bottom: 15px;"
  >
    <input
      class="form-control"
      id="expression"
      type="email"
      value="EXPRESSION"
    />
    <label
      for="expression"
    >
      Cron expression
    </label>
  </div>
</div>
`;
